### DISCUSSION TITLE
How to debug User Extension?

### DISCUSSION THREAD

**Discussion by id628** [Koenkk/zigbee2mqtt#6727] [General]
I'm looking to create a Homie protocol extension utilizing the homeassistant.js as a starting point, but my Node.js skills are a bit... rusty. 

I'm following the basic directions at https://www.zigbee2mqtt.io/information/user_extensions.html and have some code working, but unfortunately I can't seem to get vscode to utilize breakpoints in anything that loads as an extension, including the example my-first-extension.

They "set" just fine, but they never actually break, while breakpoints in homeassistant.js run just fine. I know my code is running since any logger messages I insert are running... this seems to only affect the dynamic User Extensions.

I'm using VScode 1.55.0-insider with the Auto-Attach feature on Ubuntu 20.04.2 LTS. Any suggestions?

**Comment by Koenkk** [comment]
TBH, I've never used this feature, good old `console.log()` did the trick for me.

**Reply by sjorge** [reply]
Aren’t user extensions dynamically loaded via some eval() magic? That might break the debugger if they are.

been ages since I looked at those as my 2 small personal extensions work fine. Also debugged using console.log()

**Reply by id628** [reply]
It seems that's exactly the case. Dynamically loaded extensions can't be debugged.

I've decided to just go with putting it in the lib/extension folder and updated controller.js to require it, just so I can debug it with breakpoints and variable watches to try to understand the code a bit better. I'm not sure I have the expertise anymore to do a Homie extension... but I'm giving it a shot. Time and effort will tell.

**Reply by sjorge** [reply]
@id628 just give it a shot and when you get stuck, reaching out is always helpful.
Generally Koen or someone else can help or give you pointers.

I've been doing work on z2m, herdsman, and converters for a while now and a lot of the time I do need some help or advice as this is a big complex system.

**Reply by eerison** [reply]
> TBH, I've never used this feature, good old `console.log()` did the trick for me.

I tried to use console.log, but I can see on logs :/

related with
https://github.com/Koenkk/zigbee2mqtt/discussions/23438
https://github.com/Koenkk/zigbee2mqtt/issues/22813#issuecomment-2244492861

**Comment by Christian-Me** [comment]
@id628  Hi, how far are you with your homie convention project? 
I'm using the homie convention in my DIY devices for a couple of years now (mostly ESP8266). As zigbee is becoming more popular and I more and more lazy to build all devices myself I gave it a chance with some lightbulbs, sensors and plugs - Think I like it. I was waiting for a homie extension for quite a while but it never appeared on my radar.

Just as I started to investigate if someone already did the job or is working on it before I begin to do it myself I stumbled over your post.

Chris

**Reply by id628** [reply]
@Christian-Me I wanted to do it but then life got in the way. I never got any further than the above, unfortunately.

**Reply by Christian-Me** [reply]
No problem - perhaps I come a little further, but I habe my doubts - to many entries in my to do list. Perhaps I do the translation in Node-Red but this wouldn't help other platforms supporting the homie convention
